pkg/controller/snapshot/network/controller.go

```diff
@@ -0,0 +1,375 @@
+	snaptypes "github.com/onosproject/onos-config/pkg/types/snapshot"
+	devicesnaptypes "github.com/onosproject/onos-config/pkg/types/snapshot/device"
+	networksnaptypes "github.com/onosproject/onos-config/pkg/types/snapshot/network"
+	"github.com/onosproject/onos-topo/pkg/northbound/device"
+	"time"
+)
+
+// NewController returns a new network snapshot controller
+func NewController(leadership leadershipstore.Store, networkChanges networkchangestore.Store, networkSnapshots networksnapstore.Store, deviceSnapshots devicesnapstore.Store) *controller.Controller {
+	c := controller.NewController()
+	c.Activate(&controller.LeadershipActivator{
+		Store: leadership,
+	})
+	c.Watch(&Watcher{
+		Store: networkSnapshots,
+	})
+	c.Watch(&DeviceWatcher{
+		Store: deviceSnapshots,
+	})
+	c.Reconcile(&Reconciler{
+		networkChanges:   networkChanges,
+		networkSnapshots: networkSnapshots,
+		deviceSnapshots:  deviceSnapshots,
+	})
+	return c
+}
+
+// Reconciler is a network snapshot reconciler
+type Reconciler struct {
+	networkChanges   networkchangestore.Store
+	networkSnapshots networksnapstore.Store
+	deviceSnapshots  devicesnapstore.Store
+}
+
+// Reconcile reconciles the state of a network configuration
+func (r *Reconciler) Reconcile(id types.ID) (bool, error) {
+	snapshot, err := r.networkSnapshots.Get(networksnaptypes.ID(id))
+	if err != nil {
+		return false, err
+	}
+
+	// Handle the snapshot for each phase
+	if snapshot != nil {
+		switch snapshot.Status.Phase {
+		case snaptypes.Phase_MARK:
+			return r.reconcileMark(snapshot)
+		case snaptypes.Phase_DELETE:
+			return r.reconcileDelete(snapshot)
+		}
+	}
+	return true, nil
+}
+
+// reconcileMark reconciles a snapshot in the MARK phase
+func (r *Reconciler) reconcileMark(snapshot *networksnaptypes.NetworkSnapshot) (bool, error) {
+	// Handle each possible state of the phase
+	switch snapshot.Status.State {
+	case snaptypes.State_PENDING:
+		return r.reconcilePendingMark(snapshot)
+	case snaptypes.State_RUNNING:
+		return r.reconcileRunningMark(snapshot)
+	default:
+		return true, nil
+	}
+}
+
+// reconcilePendingMark reconciles a snapshot in the PENDING state during the MARK phase
+func (r *Reconciler) reconcilePendingMark(snapshot *networksnaptypes.NetworkSnapshot) (bool, error) {
+	// Determine whether the snapshot can be applied
+	canApply, err := r.canApplySnapshot(snapshot)
+	if err != nil {
+		return false, err
+	} else if !canApply {
+		return false, nil
+	}
+
+	// If the snapshot can be applied, update the snapshot state to RUNNING
+	snapshot.Status.State = snaptypes.State_RUNNING
+	if err := r.networkSnapshots.Update(snapshot); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// canApplySnapshot returns a bool indicating whether the change can be applied
+func (r *Reconciler) canApplySnapshot(snapshot *networksnaptypes.NetworkSnapshot) (bool, error) {
+	for index := networksnaptypes.Index(0); index < snapshot.Index; index++ {
```

Comment:
Sequential ordering in the `NetworkSnapshot` store is guaranteed, so we should be able to simply check whether the prior index is complete.